source/common/network/buffered_io_socket_handle_impl.cc

```diff
@@ -0,0 +1,302 @@
+
+bool BufferedIoSocketHandleImpl::isOpen() const { return !closed_; }
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::readv(uint64_t max_length,
+                                                          Buffer::RawSlice* slices,
+                                                          uint64_t num_slice) {
+  if (!isOpen()) {
+    return {0,
+            // TODO(lambdai): Add EBADF in IoSocketError and adopt it here.
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (pending_received_data_.length() == 0) {
+    if (read_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                                 IoSocketError::deleteIoError)};
+    }
+  }
+  absl::FixedArray<iovec> iov(num_slice);
+  uint64_t bytes_offset = 0;
+  for (uint64_t i = 0; i < num_slice && bytes_offset < max_length; i++) {
+    auto bytes_to_read_in_this_slice =
+        std::min(std::min(pending_received_data_.length(), max_length) - bytes_offset,
+                 uint64_t(slices[i].len_));
+    pending_received_data_.copyOut(bytes_offset, bytes_to_read_in_this_slice, slices[i].mem_);
+    bytes_offset += bytes_to_read_in_this_slice;
+  }
+  auto bytes_read = bytes_offset;
+  ASSERT(bytes_read <= max_length);
+  pending_received_data_.drain(bytes_read);
+  ENVOY_LOG(trace, "socket {} readv {} bytes", static_cast<void*>(this), bytes_read);
+  return {bytes_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::read(Buffer::Instance& buffer,
+                                                         uint64_t max_length) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  if (pending_received_data_.length() == 0) {
+    if (read_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(IoSocketError::getIoSocketEagainInstance(),
+                                 IoSocketError::deleteIoError)};
+    }
+  }
+  // TODO(lambdai): Move at slice boundary to move to reduce the copy.
+  uint64_t max_bytes_to_read = std::min(max_length, pending_received_data_.length());
+  buffer.move(pending_received_data_, max_bytes_to_read);
+  return {max_bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::writev(const Buffer::RawSlice* slices,
+                                                           uint64_t num_slice) {
+  if (!isOpen()) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
+  }
+  // Closed peer.
+  if (!writable_peer_) {
+    return {0,
+            Api::IoErrorPtr(new IoSocketError(SOCKET_ERROR_INVAL), IoSocketError::deleteIoError)};
```

Comment:
What does test coverage for this branch look like?

Reply:
```
  result = io_handle_->writev(&slice, 1);
  ASSERT(!result.ok());
  ASSERT_EQ(Api::IoError::IoErrorCode::UnknownError, result.err_->getErrorCode());
```
https://github.com/envoyproxy/envoy/pull/13418/files/2e80156d45bedda0b368d195c52b30ffb062cff9#diff-efd790fd4b746dba5acd59b70caed1c3d036f609b9d76875bb78242fe0369f13R64-R90